utilities/write_batch_with_index/write_batch_with_index_internal.cc

```diff
@@ -631,116 +635,83 @@ bool WBWIIteratorImpl::MatchesKey(uint32_t cf_id, const Slice& key) {
+        "Merge operator must be set for column family");
   }
 
-  auto [logger, statistics, clock] = GetStatsLoggerAndClock(cf_opts);
-
   // `op_failure_scope` (an output parameter) is not provided (set to
   // nullptr) since a failure must be propagated regardless of its value.
   return MergeHelper::TimedFullMerge(
       merge_operator, key, MergeHelper::kNoBaseValue, context.GetOperands(),
-      logger, statistics, clock, /* update_num_ops_stats */ false, result,
+      ioptions.logger, ioptions.stats, ioptions.clock,
+      /* update_num_ops_stats */ false, result,
       /* columns */ nullptr, /* op_failure_scope */ nullptr);
 }
 
-Status WriteBatchWithIndexInternal::MergeKey(const Slice& key,
-                                             const Slice& value,
-                                             const MergeContext& context,
-                                             std::string* result) const {
+Status WriteBatchWithIndexInternal::MergeKeyWithPlainBaseValue(
+    ColumnFamilyHandle* column_family, const Slice& key, const Slice& value,
+    const MergeContext& context, std::string* result) {
   // TODO: support wide columns in WBWI
 
-  if (!column_family_) {
-    return Status::InvalidArgument("Must provide a column_family");
+  if (!column_family) {
+    return Status::InvalidArgument("Must provide a column family");
   }
 
-  const auto& cf_opts = GetCFOptions();
+  const auto& ioptions = GetImmutableOptions(column_family);
 
-  const auto* merge_operator = cf_opts.merge_operator.get();
+  const auto* merge_operator = ioptions.merge_operator.get();
   if (!merge_operator) {
     return Status::InvalidArgument(
-        "Merge_operator must be set for column_family");
+        "Merge operator must be set for column family");
   }
 
-  auto [logger, statistics, clock] = GetStatsLoggerAndClock(cf_opts);
-
   // `op_failure_scope` (an output parameter) is not provided (set to
   // nullptr) since a failure must be propagated regardless of its value.
   return MergeHelper::TimedFullMerge(
       merge_operator, key, MergeHelper::kPlainBaseValue, value,
-      context.GetOperands(), logger, statistics, clock,
+      context.GetOperands(), ioptions.logger, ioptions.stats, ioptions.clock,
       /* update_num_ops_stats */ false, result,
       /* columns */ nullptr, /* op_failure_scope */ nullptr);
 }
 
 WBWIIteratorImpl::Result WriteBatchWithIndexInternal::GetFromBatch(
-    WriteBatchWithIndex* batch, const Slice& key, MergeContext* context,
-    std::string* value, Status* s) {
+    WriteBatchWithIndex* batch, ColumnFamilyHandle* column_family,
+    const Slice& key, MergeContext* context, std::string* value, Status* s) {
   *s = Status::OK();
 
   std::unique_ptr<WBWIIteratorImpl> iter(
       static_cast_with_check<WBWIIteratorImpl>(
-          batch->NewIterator(column_family_)));
+          batch->NewIterator(column_family)));
 
   // Search the iterator for this key, and updates/merges to it.
   iter->Seek(key);
```

Comment:
I think it's an existing behavior but wondering worth changing it here.

If my understanding is correct, `column_family` can be null here. This NewIterator will seeking in the default column family because `batch->NewIterator(nullptr)` above will create the iterator on column_family_id 0.

This `seek()` will be wasted, because we will return `InvalidArgument` status later inside `MergeKeyWithPlainBaseValue/MergeKeyWithNoBaseValue` (new code) or `MergeKey` (old code)

Reply:
This is a bit of a chicken-and-egg problem: the seek is only wasted if `column_family` is not provided and there are merge operands for the key; however, we don't know this without doing the lookup first.

Note that `GetFromBatch` failing under these circumstances is actually documented behavior: https://github.com/facebook/rocksdb/blob/main/include/rocksdb/utilities/write_batch_with_index.h#L214-L220

Reply:
> This is a bit of a chicken-and-egg problem: the seek is only wasted if `column_family` is not provided and there are merge operands for the key; however, we don't know this without doing the lookup first.

Just to clarify, code-wise it is `FindLatestUpdate` that populates the `MergeContext`.